### ISSUE TITLE
Question on action slicing for the policy log-probabilities

### ISSUE THREAD

**Issue by sukhijab** [YifeiZhou02/ArCHer#18]
Hi, 

Thanks a lot for the great work and code! I have just started working with the code and noticed the following discrepancy in https://github.com/YifeiZhou02/ArCHer/blob/master/archer/models/archer_agent.py

In the get_action function, the output of the agent is generated using:

```
context_len = obs_ids['attention_mask'].size(1)
outputs = self.accelerator.unwrap_model(self.model).generate(**obs_ids,\
                                    max_new_tokens=self.max_new_tokens, do_sample=self.do_sample, temperature = self.temperature,\
                                    pad_token_id = self.tokenizer.eos_token_id).cpu()
outputs = outputs[:, context_len:]
```
Hence, we slice out the observations. However for the log-probabilities the following is used:

```
selected_prediction_probs = torch.take_along_dim(prediction_probs[:, obs_ids["attention_mask"].size(1)-1:-1],\
                                                 action_ids["input_ids"].unsqueeze(2), dim=2).squeeze(2)
```

I am not quite sure, why the slicing is done from obs_ids["attention_mask"].size(1)-1 till -1, instead of obs_ids["attention_mask"].size(1) till the end of the array, as is the case for the get_action function. Could you please clarify this? Thanks!

**Comment by YifeiZhou02** [comment]
Oh I think it was because of shift-one relation between the logits and the labels (basically the logits at the current token is the probability distribution for the next token). I think this is a standard practice in language modeling training.

**Comment by sukhijab** [comment]
Thanks! Yeah I figured it out in hindsight when my training didn't work^^.
